src/core/searcher.rs

```diff
@@ -11,8 +12,52 @@ use crate::store::StoreReader;
 use crate::DocAddress;
 use crate::Index;
 
+use std::collections::BTreeMap;
+use std::ops::Deref;
+use std::sync::Arc;
+use std::sync::Weak;
 use std::{fmt, io};
 
+/// Identifies the index generation accessed by a [Searcher].
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct SearcherIndexGeneration(BTreeMap<SegmentId, u32>);
+
+impl SearcherIndexGeneration {
+    pub(crate) fn from_segment_readers(readers: &[SegmentReader]) -> Self {
+        let mut segment_id_to_del_docs = BTreeMap::new();
+        for reader in readers {
+            segment_id_to_del_docs.insert(reader.segment_id(), reader.num_deleted_docs());
+        }
+        Self(segment_id_to_del_docs)
+    }
+
+    /// Segment IDs represented in this generation.
+    pub fn segment_ids(&self) -> impl Iterator<Item = SegmentId> + '_ {
+        self.0.keys().copied()
```

Comment:
Oh I knew cloned, but I didn't know about `.copied()`

Reply:
Yes, I guess that's the only benefit since `Clone` will also just copy for a `Copy` type.